For ScrollRequestType::CancelAnimatedScroll populate requestedDataBeforeAnimatedScroll

nmoucht opened the pull request:
#### fe116ce9694929a3ca27b8e660c04d645e22e4fe
<pre>
For ScrollRequestType::CancelAnimatedScroll populate requestedDataBeforeAnimatedScroll
<a href="https://bugs.webkit.org/show_bug.cgi?id=263344">https://bugs.webkit.org/show_bug.cgi?id=263344</a>
<a href="https://rdar.apple.com/114103570">rdar://114103570</a>

Reviewed by NOBODY (OOPS!).

When we get a ScrollRequestType::PositionUpdate (Non-Animated) then a ScrollRequestType::CancelAnimatedScroll,
we have some logic to set scrollPositionOrDelta so if we get an animated scroll after, this scroll position
could be forwarded. Now that we need to know the request type to access scrollPositionOrDelta, reuse
 requestedDataBeforeAnimatedScroll to pass forward the scrollPositionOrDelta and the type of the non animated
position update. Also, take this into account for non-animated ScrollRequestType::DeltaUpdate, and
add the scrollPositionOrDelta to the new delta update if it exists on the ScrollRequestType::CancelAnimatedScroll.

* Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.cpp:
(WebCore::RequestedScrollData::merge):
(WebCore::operator&lt;&lt;):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fe116ce9694929a3ca27b8e660c04d645e22e4fe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25273 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3814 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26528 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27384 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23186 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25545 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5531 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1247 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23398 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25515 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2826 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21808 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27962 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2509 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22748 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28859 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23058 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23104 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26692 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2461 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/748 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3814 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/22496 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2902 "Built successfully") | | [⏳ 🛠 jsc-mips ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2796 "Built successfully") | | [⏳ 🧪 jsc-mips-tests ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.cpp:
```diff
@@ -33,18 +33,33 @@ void RequestedScrollData::merge(RequestedScrollData&& other)
         // Carry over the previously requested scroll position so that we can set `requestedDataBeforeAnimatedScroll`
         // below in the case where the cancelled animated scroll is immediately followed by another animated scroll.
-        other.scrollPositionOrDelta = scrollPositionOrDelta;
+        other.requestedDataBeforeAnimatedScroll = { requestType, scrollPositionOrDelta, scrollType, clamping };
```
What if `this.requestType` is `ScrollRequestType::CancelAnimatedScroll`? Two cancels can be merged.

nmoucht replied:
Right changed this to forward requestedDataBeforeAnimatedScroll.

smfr commented on Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.cpp:
```diff
@@ -33,18 +33,33 @@ void RequestedScrollData::merge(RequestedScrollData&& other)
     } else if ((other.requestType == ScrollRequestType::PositionUpdate || other.requestType == ScrollRequestType::DeltaUpdate) && other.animated == ScrollIsAnimated::Yes) {
         switch (animated) {
-        case ScrollIsAnimated::No:
-            other.requestedDataBeforeAnimatedScroll = { requestType, scrollPositionOrDelta, scrollType, clamping };
```
Now that I read this code, I'm confused how we set `other.requestedDataBeforeAnimatedScroll`. Aren't we merging `other` into `this`?

Also, should we assert that `other` has `requestedDataBeforeAnimatedScroll`, because if it does that makes everything way more complicated.

This makes me wonder if trying to do this merging here is fraught with peril; maybe we should just store a vector of `RequestedScrollData` and work out what happens by looking at the entire list at handling time.

nmoucht replied:
Yeah its kind of strange but we modify other then do` *this = WTFMove(other);`.

smfr replied:
I think we should flip this around so I'm not confused in future. Logically, we're appending `other` to `this`.

smfr replied:
The way I'd write this function is to switch on `requestType` and inside each clause switch on `other.requestType`. If there's shared logic between branches, express it via lambdas. And just modify `this`, not `other`. Also assert that `other` has no `requestedDataBeforeAnimatedScroll`.

nmoucht replied:
Updated to this implementation style. The code is simpler if we merge this -> other vs the other way around. I changed to creating a temporary RequestScrollData to be the merge result, which should hopefully help with readability. Added assert statement.

nmoucht committed:
For ScrollRequestType::CancelAnimatedScroll populate requestedDataBeforeAnimatedScroll
https://bugs.webkit.org/show_bug.cgi?id=263344
rdar://114103570

Reviewed by NOBODY (OOPS!).

When we get a ScrollRequestType::PositionUpdate (Non-Animated) then a ScrollRequestType::CancelAnimatedScroll,
we have some logic to set scrollPositionOrDelta so if we get an animated scroll after, this scroll position
could be forwarded. Now that we need to know the request type to access scrollPositionOrDelta, reuse
 requestedDataBeforeAnimatedScroll to pass forward the scrollPositionOrDelta and the type of the non animated
position update. Also, take this into account for non-animated ScrollRequestType::DeltaUpdate, and
add the scrollPositionOrDelta to the new delta update if it exists on the ScrollRequestType::CancelAnimatedScroll.

* Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.cpp:
(WebCore::RequestedScrollData::merge):
(WebCore::operator<<):
